### DISCUSSION TITLE
Display to user from system script?

### DISCUSSION THREAD

**Discussion by DamagedDingo** [swiftDialog/swiftDialog#379] [Q&A] [answered]
I'm fairly new the whole macOS thing and swiftDialog only today so please forgive me if this is a dumb question. 

Can I run a script from intune as system and have it display the dialog to the signed in user? 
I tried this which was the only thing that I could find that might be right. 
```zsh
loggedInUser=$(stat -f %Su /dev/console)
sudo -u $loggedInUser dialog --message "etc."
```

**Comment by bartreardon** [comment, marked as answer]
[`/usr/local/bin/dialog`](https://github.com/swiftDialog/swiftDialog/blob/main/dialog/Package/pkgroot/usr/local/bin/dialog) will do that for you if it needs to. You can run that as the user or from a system process and it will launch in the correct context.

**Reply by DamagedDingo** [reply]
Argh that was so simple... I thought it was in Path so was I only using dialog but with full path its working now. 
Thanks for the super fast reply!

**Reply by bartreardon** [reply]
noproblemo 🙂

**Reply by DamagedDingo** [reply]
I still have something wrong. 

```zsh
/usr/local/bin/dialog \
    --title "Updating GlobalProtect VPN Client" \
    --message "Please wait while we update GlobalProtect.<br><br>" \
    --commandfile $swiftDialogCommandFile \
    --icon "https://url.com/image.jpg"
```

But it sits on this until I click continue, which closes the dialog. I want to be able to update it using the CommandFile but it won't continue to the next line in the script where I update it.
```zsh
sudo sh -c "echo 'progress: 05' >> $swiftDialogCommandFile"
sudo sh -c "echo 'progress: show' >> $swiftDialogCommandFile"
```
I also tried --button1disabled but then the dialog shows with no way to exit.

**Reply by BigMacAdmin** [reply]
@DamagedDingo By default, shell scripts run one command at a time and do not begin the next until the previous has completed.

You need to tell your script "Open a dialog window, and send that process to the background so we can continue" and the way to do that is with an `&` right at the end of your command.
```bash
/usr/local/bin/dialog \
    --title "Updating GlobalProtect VPN Client" \
    --message "Please wait while we update GlobalProtect.<br><br>" \
    --commandfile $swiftDialogCommandFile \
    --icon "https://url.com/image.jpg" &
```

Here is an example script of how to send updates to a running dialog (but it does look like you've got it 95% of the way there already): https://github.com/SecondSonConsulting/swiftDialogExamples/blob/main/progressExample.sh

Also, if you are not yet a member of the MacAdmins Slack, it's free and there is a channel dedicated to SwiftDialog with almost 2,000 people that can help with questions like this. You'll get more visibility and faster answers than asking here on Github where Bart may be the only one to notice. You can join here: https://macadmins.org

**Reply by BigMacAdmin** [reply]
You also almost certainly don't need to use `sudo` in your scripts. If you're running from a management system like Intune, the entire script will already be running with root privileges and sudo isn't needed.

`sh -c "echo 'progress: 05' >> $swiftDialogCommandFile"` is also a bit of a sideways way to do what you're doing as well. `sh -c` says "Execute the following string in `sh`" but if you're running from Intune, it's already running (probably as bash not sh) and you can just use `/bin/echo 'progress: 05' >> $swiftDialogCommandFile"` without the `sh -c`

**Reply by DamagedDingo** [reply]
Thank you. I was looking for a sample all day but couldn't find anything. 
I'll check out slack too.

**Reply by DamagedDingo** [reply]
I should have removed that. I thought it was failing writing to the file at first and that was why it was not continuing, but i added logging after that and realised that it just wanted going past the first command. 
I'll clean all that mess up now.

**Reply by bartreardon** [reply]
I have a bunch of examples here as well https://github.com/bartreardon/swiftDialog-scripts/ which might help. (Some more out of date than others but all should work still)
